Add Equation Class to Sympy Core

gutow opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
This pull request replaces #21333. Inclusion of this class in sympy was most recently mentioned in this [discussion thread](https://groups.google.com/g/sympy/c/GB5pIUjS1as/m/M9NQjQMUAgAJ). I apologize for the delay in creating a related pull request, but have been very busy.

#### Brief description of what is fixed or changed

The pull request adds the core concept of an equation with the two sides
connected by an equals (=) sign, that can be algebraically manipulated by performing operations
on the equation. Unlike, the `Eq` class this does not collapse to `True` or `False` if Sympy can determine the logical relation between the two sides. As the Boolean behavior of `Eq` is required for some other parts of Sympy a new class is required to avoid the collapse.

Because `Equation` defaults to performing the same operation on both sides simultaneously,
it provides a simple way to input steps in an algebraic manipulation. Once Equation is included the `solve()` operation can be modified to accept Equations as input.

#### Other comments

- The changes have been made in a way that changes none of the behavior of the rest of Sympy.
- In an environment that produces typeset output (Jupyter/IPython) equations appear with an '=' sign. As normal the executable version can be got using `srep()`.
- The Equation class has been maintained and used for computer assisted algebra with students as part of the extension called [Algebra_with_Sympy](https://gutow.github.io/Algebra_with_Sympy/algebra_with_sympy.html) since 2022 (about 3 yrs).
- It was found that because of the way sympy strictly calls internal versions of operations, certain operations (eg. 'simplyfy()`) would return incompatible objects. Thus for the past 1.5 years the Equation class has been maintained and used for computer assisted algebra with students in a fork of sympy ([sympy_for_algebra](https://github.com/gutow/sympy-for-algebra)) with good success.
- User interface and convenience tools that facilitate the use of this class for algebraic manipulations are maintained in a separate extension  [Algebra_with_Sympy](https://gutow.github.io/Algebra_with_Sympy/algebra_with_sympy.html).
- Tests include automatic testing against all registered functions in Sympy. This facilitates tracking compatibility with newly added functions.
- This implementation provides convenience methods previously requested by others: `.swap` to swap the sides, `.as_Boolean() ` to convert to type `Eq`, `.check()` as a shortcut for `.as_Boolean().simplify()`.,  `.do.`, `.dolhs.`, `dorhs.` ,`.apply.`, `.applylhs.`, `.applyrhs.` .

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * Adds the `Equation` (short name `Eqn`) class that behaves like as an algebraic equation with the relational operator `=`.
  * Adds tests for the basic algebraic behavior of the `Equation` class.
  * Adds tests of all registered functions for compatibility with the `Equation` class.

* other
  * In `setup.py` added equation to modules and tests.
  * Added imports of equation module to `sympy.__init__.py`.
  * Added imports of equation module to `sympy.core.__init__.py`.

* functions
  * modified `function` class to recognize equations.

* simplify
  * modified `radsimp.py` to also handle equations by tweaking the `_eval_collect` operation.

* printing
  * modified printing to recognize equations

<!-- END RELEASE NOTES -->

gutow committed:
Basic Equation functionality re-embedded

gutow committed:
extention of sympy operations to Equations

gutow committed:
Equation tests

gutow committed:
Readme Updates

gutow committed:
flake8 fixes

gutow committed:
removal of algwsym_config references

gutow committed:
fix usage of importlib

gutow committed:
added Equation to test_args.py

gutow committed:
Remove commented out code

gutow committed:
flake8 list comprehension fix

gutow committed:
rounding fix in equation tests, udivisor_sigma added to sympy init imports.

gutow committed:
try to fix mystery not used error

gutow committed:
Revert README.md to default sympy

gutow committed:
remove finding eqn python name

gutow committed:
test_equation.py updates

gutow committed:
move as much to sympy.printing as possible.

gutow committed:
latex test.

gutow committed:
printing and test updates

gutow committed:
fix whitespace issues

gutow committed:
add newline at end of test_equation.py

gutow committed:
trying to make equation class more standard Sympy

gutow committed:
removing implicit import in equation.py

gutow committed:
force Ruff to ignore Base Level Exception error in testing for generic failures of sympy objects as part of equations

gutow committed:
uncommenting __repr__ override.

gutow committed:
make srepr work as normal sympy.

gutow committed:
doctests to match sympy output overrides

gutow committed:
Import Equation into function.py earlier

gutow committed:
remove unneeded Eqn import in function.py

gutow committed:
reinsert missing unit definitions in equation.py doctests

gutow committed:
extend function types to include equations

gutow committed:
output helpers test to sympy standards

gutow committed:
set str.py back to executable output and add pretty_print

gutow committed:
doctests and test back to `Eqn(lhs,rhs)` form

gutow committed:
printing tests

gutow committed:
docstring fixes

gutow committed:
try allowing "FLOAT_CMP" in .toml file in doctest_optionflags

gutow commented:
I do not see a reason to preserve the history here. It should be squashed before merging.

I maintained as much history as possible in my working repository, to help me get a sense of how Sympy changes were interacting with the Equation class. The good news is other than the introduction of new functions, which is tested for, I have seen zero issues with other changes.

sylee957 commented on sympy/simplify/radsimp.py:
```diff
@@ -169,6 +169,10 @@ def collect(expr, syms, func=None, evaluate=None, exact=False, distribute_order_
 
     collect_const, collect_sqrt, rcollect
     """
+    _eval_collect = getattr(expr, '_eval_collect', None)
+    if _eval_collect is not None:
+        return _eval_collect(syms, func, evaluate,
+                             exact, distribute_order_term)
```
Where did this change came from?
I don't think that we are using `_eval_collect` anyway
https://github.com/search?q=repo%3Asympy%2Fsympy%20_eval_collect&type=code

gutow replied:
Is this resolved for now?

asmeurer commented:
It's going to be important that we have good documentation here on what Equation is and how it differs from Eq. Both classes should clearly refer to each other.

asmeurer commented:
Are you planning to integrate this with `solve()`?

oscarbenjamin commented:
> One issue I noticed is that this doesn't work very well if you try to put an Equation directly in one of the core classes:
> ```python
> >>> Mul(Equation(x, y), z)
> ```

Why would you want to do that?

> It's not clear to me that `Equation` should be `Basic` and not `Expr`

It should not be `Expr` precisely because it should not be in places like the args of a Mul.

gutow commented on sympy/core/equation.py:
```diff
@@ -0,0 +1,638 @@
+                        self.rhs.diff(*args, **kwargs))
+
+    def _eval_Integral(self, *args, **kwargs):
+        side = kwargs.pop('side', None)  # Could not seem to pass values for
```
I will experiment again, but it did not work the last time I tried. `diff()` was not passing on kwargs it did not recognize. Maybe it has been fixed.

gutow replied:
for `Equation._eval_Integral()` the keyword 'evaluate' is removed before reaching this operation. The `side` keyword reaches this operation without a problem. That is what the comment is about. Maybe the comment should be a "`TODO: look into getting evaluate keyword to pass through to here`"?  I do not think the current behavior causes serious issues although if one wants an equation that looks like an integral table entry the two sides have to be handled separately, as in the doctest starting at line 230 of equation.py:
```
    Make a pretty statement of integration from an equation
    >>> Eqn(Integral(q.lhs,b),integrate(q,b,side='rhs'))
    Eqn(Integral(a*c, b), b**2/(2*c))
```

gutow replied:
Is this resolved for now?

asmeurer replied:
I'm not sure if I'm misunderstanding what the issue is here, but my suggestion was just to use

```py
def _eval_integral(self, *args, side=None, **kwargs):
```

instead of using `kwargs.pop('side', None)`. 

In general, `kwargs.pop` can be replaced with an explicit keyword argument like this which is a little easier to read. 

In some cases, you can completely replace `**kwargs` if you do this, but in this case, it sounds like you'll have to leave it to handle `evaluate`.

gutow commented:
> Are you planning to integrate this with `solve()`?

The intent is to do this after `Equation` is merged. I have it working in the Algebra_with_Sympy extension. As the changes go into `solve()` and involve making some choice about the output format, I thought it should be a separate pull request.

I will update the description to make this clearer.

oscarbenjamin commented:
> > Are you planning to integrate this with `solve()`?
> 
> The intent is to do this after `Equation` is merged.

I agree with leaving things keeping things simple for now until this PR is resolved but my 2 pence is that I would like `Equation` to have a method called something like `solve_for` or `make_subject` that rearranges the equation to make one variable the subject. It should not call anything like `solve` but only do limited invertible manipulations e.g.:
```python
In [2]: eq
Out[2]: 2⋅x + 1 = y

In [3]: eq.make_subject(y)
Out[2]: y = 2⋅x + 1

In [3]: eq.make_subject(x)
Out[2]: x = (y - 1)/2
```

oscarbenjamin commented:
As for `solve` there is a bit of a can of worms here. If I was rewriting solve from scratch I would have a "system of equations" type that represents a whole system of equations, unequalities and inequalities and that has methods like `factor`, `clear_denominators`, `reduce`, `eliminate` etc for manipulating the system as a whole. The `solve` function would then use that type internally rather than just as an input form.

gutow commented:
I agree that `solve()` is quite a can of worms, but it is not hard to have it accept an Equation or system of Equations and reformulate each equation as `0=Expr`, then use the rhs as normal in `solve()`. This immediately provides current functionality to Equations.

I do like your concept for a re-implementation of `solve()`. Could that be pursued as a reworking of `SolveSet()`, which I think is headed that way?

I think I do not understand what simple implementation you are suggesting within the Equation class.

gutow commented:
> In general I am in favour of the idea although I'm interested to see if anyone else has views.
> 
> I will refrain from detailed review for now but one thing I will say is that it looks like something funny has happened with the git history (see the note above about commits deleting files). Is there a reason to preserve the history here rather than just squashing all the commits?

Squashing is not something I have any significant experience with, so may need some help. Is is possible for whomever has merge permission to squash when they do the merge?

oscarbenjamin commented:
> Is is possible for whomever has merge permission to squash when they do the merge?

Not easily. See gh-18548.

gutow commented:
> Another issue I've noticed is that we already have unevaluated expressions, but Equation doesn’t seem to respect them consistently. For example:
> 
> ```
> sin(pi / 3, evaluate=False)
> # sin(pi/3)
> sin(Eqn(x, y), evaluate=False)
> # Eqn(sin(x), sin(y))
> ```

It works as I expect:
```
>>> sin(Eqn(pi/3,I))
Eqn(sqrt(3)/2, I*sinh(1))
>>> sin(Eqn(pi/3,I), evaluate=False)
Eqn(sin(pi/3), sin(I))
```
Were you expecting `sin(Eqn(x,y)) -> sin(Eqn(x,y))`? I interpreted the expected behavior to be that `evaluate = False` would be passed on to the operations applied to each side.

gutow committed:
fixed over indented test in test_equation.py

sylee957 commented:
Furthermore, even if `sin(x = y)` is implemented, I would like to discourage fellow SymPy developers to write internal solvers based on that because we will see people mixing expressions and equations everywhere.

gutow commented:
> I would not have made it possible to call a function on an equation like this. For one thing it creates an expectation that every part of the codebase would have to check for Equation.

I am not sure what you are referring to? As it presently stands no part of the codebase except for the base function class has to be aware of the Equation class to get this behavior. This behavior is implemented by simple changes to the base class for function, that distributes the operation of the function to the expr on each side. See the changes to functions.py starting at line 454. This is just like applying it to a two element vector. I have actually tried implementing these ideas on top of vectors, but the present implementation is slimmer and easier to handle with the output/printing engine of sympy.

There are some functions where things do not work well because it is not clear how to apply the operations to both sides of an equation. These are listed in the object `_not_applicable_to_equations_` in test_equation.py. This short list of functions  contains things such as `interpolating_spline` which requires a specific data set form on which to operate. In theory, one could build an expression on both sides of the equation that would work, but I am not sure it would be particularly useful. That said, I don't think we should disallow it because in that case a valid statement could be generated.

When a new function is added to sympy the test code will pick it up and throw an error if `f(eqn) != Eqn(f(eqn.lhs),f(eqn.rhs))` or an error is thrown because the signature cannot be used to generate valid input to the function. The special checks for 'betainc'  and 'betainc_regularized' at line 39 of test_equation.py are the two cases where I was able to figure out what was not specified properly by the signatures.

The functions/operations in `_not_applicable_to_equations_` were initially identified by running the test code with `_not_applicable_to_equations_` empty. After the initial run the only time I've seen a failure was when a bug was introduced into the Heaviside function. The bug was fixed and the problem went away. test_equations.py still explicitly tests the Heaviside function  at line 214, because I had an override (now removed) to fix the bug.

The intent is to automatically identify functions added to sympy that do not work with this paradigm and deal with them within the equation class if necessary, or ignore if the operation does not make sense in an equation. All cases that do not work throw an error for the user, just as they do when the operation fails on an expression.

oscarbenjamin commented:
> As it presently stands no part of the codebase except for the base function class has to be aware of the Equation class to get this behavior.

I don't like having this special case there. It makes the signature ambiguous. You have changed the return type annotation for `DefinedFunction.__new__` from `-> Expr` to:
```
    def __new__(cls, *args, **options) -> Expr | Equation: 
        return cls._new_(*args, **options)
```
Firstly this should really use `@overload` so that a type checker can tell when it would return `Equation` vs when it would return `Expr`. Having just `Expr | Equation` here would reek havoc on type inference and checking. The issue is not really about static type annotations though but rather that the ambiguous return type is a problem in itself. Right now you can trust that `sin(obj)` will return `Expr` without checking the type of `obj` but the special case here would change that.

Users also don't generally know the difference between a symbolic function like `sin` vs a Python function like `sqrt` or `factor`. If they see that you can do `sin(eqn)` then they will expect to be able to pass the Equation type everywhere. Of course some operations like `factor` should be defined for `Equation` but there are also many other functions that cannot be expected to check for `Equation`. Instead if `Equation` works with functions like `factor` then it should be because `factor` has a general way of working with `Basic` rather than a special case check for the Equation type.

Having this kind of special case in `DefinedFunction.__new__` affects many classes because there are so many subclasses. The code in `function.py` is already a gnarly mess and we should not add anything there that special cases particular types. It might seem like a "small" change because it is only a small amount of code but it is a very far reaching change in terms of effect and also the precedent that it sets.

Not all expression types in sympy are subclasses of DefinedFunction. There are also things like Piecewise or Sets and so on so just patching something into `DefinedFunction.__new__` does not cover all the cases where the same behaviour might be expected.

gutow committed:
Comment updates for clarity

asmeurer commented:
I'm tending to agree that functions applying to equations could be problematic. How important is that functionality for the usability here? 

One thing that also concerns me is that this is mostly a syntactic sugar to minimize typing. SymPy usually tries to mimic real world mathematical syntax and conventions. But $$(x = y)*z$$ or $$\sin(x = y)$$ is not something that people usually write. You would only ever really see the former in the "work" from an algebra student, but it would generally not be considered rigorous notation. So given this, I think it's a mistake if this requires too many modifications directly to the core, because the core should really only be aware of things that affect expressions that are used generally.

oscarbenjamin commented:
> SymPy usually tries to mimic real world mathematical syntax and conventions. But 
>  is not something that people usually write.

This is because if you are writing it out manually then you still have to write out both sides of the equation manually anyway. Here you are getting the computer to remember the sides and do the manipulations for you and you want to tell it what to do succinctly:
```
>>> eq = Equation((x - 1)/2, y)
>>> 2*eq + 1
Eq(x, 2*y + 1)
```
The advantage is that the equation type ensures that you are applying the same operations to both sides. I would expect this sort of thing to be used interactively or in a notebook or something rather than in internal code in SymPy.

The alternatives are:
```
>>> eq.apply(lambda s: 2*s + 1)
>>> Equation(2*eq.lhs + 1, 2*eq.rhs + 1)
```
I would say that the alternative forms are needed for functions like `sin` anyway though.

oscarbenjamin commented:
One thing you do see people write is something a bit like this:
```
(x - 1)/2 = y
x - 1 = 2*y     (x 2)
x = 2*y + 1     (+ 1)
```
That is the main statement is always an equation but you have little notes like that explain what manipulation is being applied to both sides to get from one equation to another. The idea here is that you tell the computer what you want to do `(x 2)` and it does that to both sides.

gutow committed:
fix whitespace in test_args.py

gutow commented:
I believe the last two commits clean up the Git History as much as I can. It still is not great. The only way I could get it to work was  to cherry-pick into a new branch. This lost some author attributions for @Davide-sd and @oscarbenjamin.

asmeurer commented:
I think the main thing I'd like to see here still is some documentation going over when you'd use Eq and when you'd use Equation, and what the design decisions are for each. For example, this is a pretty fundamental difference:

```py
>>> Equation(0, 1)
Eqn(0, 1)
>>> Eq(0, 1)
False
```

It would probably be best to write this as a rst or md document in the docs, and reference it in the Equation and Eq docstrings. Also Equation itself could be added to the API docs.

gutow committed:
note who suggested or provided some code

Covers suggestions and Git attributions lost in transfer from Algebra_with_Sympy to Sympy.

gutow committed:
whitespace fixes in docstrings for Pyodide

gutow committed:
Remove Unicode in name

gutow commented:
Question related to deprecation of behavior in the `Equality` class that is also implemented in the `Equation` class.

I noticed a deprecation for rewrite in the `Eq` class. The same behavior that is being deprecated in `Eq` is implemented for the `Equation` class in this pull request. The Deprecation is:

**Deprecate Eq.rewrite(Add)**
_The ability to rewrite eq = Eq(x, y) like eq.rewrite(Add) to give x - y has been deprecated in favor of writing eq.lhs - eq.rhs. A replacement property/method was not deemed necessary given the clarity of the explicit use of lhs and rhs, and the inclusion of this functionality in the rewrite apparatus leads to failures when a node expecting a Boolean is re- written as an Expr._

Since `Equation` will not collapse to a Boolean, the problem this deprecation avoids is not an issue. Thus I would argue that the behavior should remain in the `Equation` class. It was implemented because people requested it.

Thoughts?

Firionus commented:
As the one who first observed the bug (https://github.com/gutow/Algebra_with_Sympy/issues/23) that showed there need to be changes made to core sympy for `Equation`, I'd like to reinforce that there is a real user need for this functionality. 
As a physics student, I've been taught to perform math manipulations on equations, not expressions. Thus, trying to check or do my homework with sympy was always problematic, since sympy requires reframing all of that in terms of expressions. `Algebra_with_Sympy` and `Equation` was a breath of fresh air for me, since I could finally start interacting with the CAS in a way that math would be manipulated on paper or in publications, allowing me to "step-through" my math like in a "math debugger". 
I personally don't care whether this ability is part of core sympy or part of a package. However, you might consider that if sympy becomes more usable to students for everyday manipulation, this might increase its usage in research and industry later on. 

When it comes the API, I want to say that the suggested form 
```
>>> eq = Equation((x - 1)/2, y)
>>> eq.apply(lambda s: 2*s + 1)
>>> Equation(2*eq.lhs + 1, 2*eq.rhs + 1)
```
seems extremely unergonomic to me, defeating the purpose of offering an easy way to interact with on-paper math.

This sort of natural API is a must for me with +,-,*,...:
```
>>> eq = Equation((x - 1)/2, y)
>>> 2*eq + 1
Eq(x, 2*y + 1)
```
And staying with that paradigm, it makes a lot of sense to offer something like
```
>>> eq = Equation(2**x, 10)
>>> log(eq)
Eq(log(2**x), log(10))
```
to lead towards `x=log(10)/log(2)`. It seems that enabling this behavior is the only problem here and I hope it can be resolved.

In case it can't, I personally don't hate the variant `eq.apply(log)` suggested by @oscarbenjamin, which would allow algebra-with-sympy to stay separate, as far as I understand.

sylee957 commented:
> As the one who first observed the bug (https://github.com/gutow/Algebra_with_Sympy/issues/23) that showed there need to be changes made to core sympy for Equation, I'd like to reinforce that there is a real user need for this functionality.

I appreciate your efforts. However, you should know that SymPy is a very large library and it is nearly impossible for anyone to draw a full picture of core usage, all combinations of equations, and compatibility issues. If I am working with an unclear codebase, I would always prioritize about making things maximally modular. I do not like the attitude of always trying to make core changes and prove that nothing breaks by just adding more test cases.

sylee957 commented:
I also want to raise some fundamental questions that have gone unanswered for years when discussing equations in SymPy.

Examples:

- Can you divide both sides of an equation by x? How is the x = 0 case handled?
- Many functions in sympy are multivalued or conditional. Applying sqrt to x^2 = n yields only one branch; what about the other?

Frankly, I don’t care how many “use cases” you list. The issue is that you leave these questions unanswered, and they aren’t clarified in the documentation. One could set out an axiomatic basis (reflexivity, symmetry, transitivity; clear definitions of terms/substitution) and specify when function application is valid (including required case splits), or explicitly state that proofs about function domains are left to users. I haven’t seen that work from you.

A clear formal specification is good because it gives more implementation freedom. I’m fine with allowing function application—if there’s a formal reference stating when it is correct. However, coding first and expecting others to follow later doesn’t fly. When bugs appear, we don’t even know what to fix: add conditional checks, or not? (This is also why we keep insisting on keeping the code simple: if the formal specification isn’t clear, the code should be simple enough that its intended semantics are obvious. If you can clarify the spec, this point becomes moot.) If developer A assumes sqrt is the principal branch while developer B treats it as multivalued, communication breaks down.

Yes, high-school algebra is not rigorous; that’s exactly why you should define these points clearly. Formal definitions help communication and alignment.

If you can’t settle these points, let’s invite a trained logician to pin down the formal meaning of equations and the core algebraic operations.
